### DISCUSSION TITLE
Need a nicer way of building plugins by assembling others.

### DISCUSSION THREAD

**Discussion by jonapgar-groupby** [dotansimha/envelop#1818] [General]
Consider this "AuthPlugin" I've made here:

```typescript
const authPluginFactory = (
  options: ReturnType<typeof buildAuthConfig>,
  prisma: PrismaClient
): Plugin<Pick<Context, 'agent'>> => ({
  onPluginInit({ addPlugin }) {
    // addPlugin should not require the added plugin to satisfy the context, but it does :shrug:
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    addPlugin(useAuth0(options) as Plugin<any>);
    addPlugin(useExtendContext(createAgentContextExtender(prisma)));
  },
});
```

Now I know I could have just done this:

```typescript
plugins:[
  useAuth0(options),
  useExtendContext(createAgentContextExtender(prisma)
],
```

But I have ioc scenarios (e.g. testing) where I don't want to "useAuth0" at all, and I want my AuthPlugin provider to deal with all of that.

The above solution with calling`addPlugin` multiple times works but as you can see I have to assert `Plugin<any>`, as addPlugin expects that I'm passing a Plugin that satisfies the PluginContext of the plugin that I'm actually creating.

It would be great if I could just call a method, something like this...

```typescript
const combinePlugins<T1, T2, T3>(Plugin<T1>, Plugin<T2>, Plugin<T3>): Plugin<T1 & T2 & T3> = (...
```

Though maybe this doesn't make a lot of sense either. 

The whole idea of PluginContext is kinda confusing to me. It seems like this type is doing double-duty and in some places is implying that the Plugin will extend the context with these fields, and other places it is implying "the context already has these fields".

The `AfterContextBuildingEventPayload` shows this strange duality.

```
/**
 * The payload forwarded to the afterContextBuilding hook.
 */
export type AfterContextBuildingEventPayload<ContextType> = {
  /**
   * The current context object.
   */
  context: ContextType;
  /**
   * Extend the context object with a partial.
   */
  extendContext: (contextExtension: Partial<ContextType>) => void;
};

**Comment by jonapgar-groupby** [comment]
I guess ideally the plugins array in `envelop` would enforce that the _in_ and _out_ context for each plugin. Plugins could require that a previous plugin as added some field to the context, and they can decide to extend/remove that field in their own implementation.

**Comment by n1ru4l** [comment]
Related https://github.com/n1ru4l/envelop/issues/1566
